kernel/src/util/iovec.rs

```diff
@@ -37,92 +39,190 @@ impl TryFrom<UserIoVec> for IoVec {
-        Ok(len)
+    pub fn writers_mut(&'a mut self) -> &'a mut [VmWriter<'a>] {
+        &mut self.0
     }
+}
 
-    /// Writes bytes from the `src` buffer
-    /// to the user space buffer pointed by the `IoVec`.
+/// Trait defining the read behavior for a collection of [`VmReader`]s.
+pub trait IoVecRead {
+    /// Reads data to the provided writer from all collected [`VmReader`]s,
+    /// accumulating total bytes read.
     ///
-    /// If successful, returns the length of actually written bytes.
-    pub fn write_to_user(&self, src: &[u8]) -> Result<usize> {
-        let len = self.len.min(src.len());
-        CurrentUserSpace::get().write_bytes(self.base, &mut VmReader::from(&src[..len]))?;
-        Ok(len)
+    /// This method returns [`Errno::EFAULT`]
+    /// if a page fault occurs and _not all bytes_ are successfully read.
+    fn read_exact(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize>;
+
+    /// Returns the length of data remaining to read.
+    fn len(&self) -> usize;
+
+    /// Checks if the data remaining to read is empty.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
     }
 }
 
-/// Copies IO vectors from user space.
-pub fn copy_iovs_from_user(start_addr: Vaddr, count: usize) -> Result<Box<[IoVec]>> {
-    let mut io_vecs = Vec::with_capacity(count);
+/// Trait defining the write behavior for a collection of [`VmWriter`]s.
+pub trait IoVecWrite {
+    /// Writes data from the provided reader to all collected [`VmWriter`]s,
+    /// accumulating total bytes written.
+    ///     
+    /// This method returns [`Errno::EFAULT`]
+    /// if a page fault occurs and _not all bytes_ are successfully written.
+    fn write_all(&mut self, reader: &mut VmReader<'_, Infallible>) -> Result<usize>;
+
+    /// Returns the length of space available to write.
+    fn len(&self) -> usize;
+
+    /// Checks if the space available to write is empty.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
 
-    let user_space = CurrentUserSpace::get();
-    for idx in 0..count {
-        let addr = start_addr + idx * core::mem::size_of::<UserIoVec>();
-        let uiov = user_space.read_val::<UserIoVec>(addr)?;
-        let iov = IoVec::try_from(uiov)?;
-        io_vecs.push(iov);
+impl<'a> IoVecRead for IoVecReader<'a> {
+    fn read_exact(&mut self, writer: &mut VmWriter<'_, Infallible>) -> Result<usize> {
+        let mut total_len = 0;
+
+        for reader in &mut self.0 {
+            match reader.read_fallible(writer) {
+                Ok(copied_len) => total_len += copied_len,
+                Err(_) => return_errno_with_message!(
+                    Errno::EFAULT,
+                    "cannot copy from `IoVecReader` to `VmWriter`"
+                ),
+            }
+        }
+        Ok(total_len)
+    }
+
+    fn len(&self) -> usize {
+        self.0.iter().map(|vm_reader| vm_reader.remain()).sum()
```

Comment:
Do not name a method as if it is a simple getter when the method's implementation involves a loop. The caller would be given the false impression that the method is cheap, but it is not that cheap.

Reply:
This method is renamed as `calculate_total_length`.

Reply:
For brevity, how about `sum_lens`?